### PR TITLE
Fix #1173: preferences' values mutability

### LINKED ISSUE #1173: Preferences do not return a copy of the value
i.e 
```typescript
methodThatIsCalledSeveralTimes() {
    /* Here we get the preferences from the injected pref service */
    const prefValues: string[] = this.PrefService['testKey'];
    /* We can then modify that preference by adding stuff to it, and this would directly edit the array in the 
       service's preference cache, as it returned a reference in the first case */
    prefValues.push('something');
    /* Next time the function is called, prefValues will have 'something' in it (plus the preferences that were there before) */
}
```

### PR DESCRIPTION
**Pull request by epatpol**
Example of a fix for #1173, wdyt @akosyakov ? I think even the configuration i.e defaults could have an array of preferences thus returning a reference, couldn't it?

### CODE REVIEW

File: packages/preferences-api/src/common/preference-proxy.ts [hunk 1]
```diff
@@ -55,16 +56,16 @@ export function createPreferenceProxy<T extends Configuration>(preferences: Pref
         get: (_, p: string) => {
             if (p in schema.properties) {
                 if (p in configuration) {
-                    const preference = preferences.get(p, configuration[p]);
+                    const preference = clone(preferences.get(p, configuration[p]));
```
Comment (akosyakov) [inline packages/preferences-api/src/common/preference-proxy.ts]: proxies are just convenient interfaces, functionality should be implemented in the lower level in `PreferenceService`

File: packages/preferences-api/package.json [hunk 2]
```diff
@@ -5,7 +5,9 @@
   "main": "lib/common/index.js",
   "typings": "lib/common/index.d.ts",
   "dependencies": {
-    "@theia/core": "^0.3.3"
+    "@theia/core": "^0.3.3",
+    "@types/clone": "^0.1.30",
```
Comment (akosyakov) [inline packages/preferences-api/package.json]: I would prefer to avoid adding new dependencies if one can achieve results with existing dependencies, like with phosphor

**Comment by epatpol** [comment]
I'd like to merge #1158 before adding this one as it already changes quite a bunch of stuff

**Comment by simark** [comment]
Maybe you intend to do it after, but can you give your commit a more meaningful name?

**Comment by simark** [comment]
It seems to fix my use case.

**Comment by epatpol** [comment]
Good point, I'll edit the commit message to something more meaningful

File: packages/preferences-api/src/common/preference-service.ts [hunk 3]
```diff
@@ -65,7 +66,7 @@ export class PreferenceService implements Disposable {
     get<T>(preferenceName: string, defaultValue: T): T;
     get<T>(preferenceName: string, defaultValue?: T): T | undefined {
         const value = this.prefCache[preferenceName];
-        return value !== null && value !== undefined ? value : defaultValue;
+        return value !== null && value !== undefined ? clone(value) : clone(defaultValue);
```
Comment (akosyakov) [inline packages/preferences-api/src/common/preference-service.ts]: looking at https://github.com/theia-ide/theia/blob/master/packages/preferences-api/src/browser/preference-proxy.ts#L65 again, not all values returned by a preference proxy come from here, we have to clone there as well

File: packages/preferences-api/src/common/preference-service.ts [hunk 4]
```diff
@@ -49,7 +50,7 @@ export class PreferenceService implements Disposable {
 
         this.resolveReady();
         for (const change of event.changes) {
-            this.onPreferenceChangedEmitter.fire(change);
+            this.onPreferenceChangedEmitter.fire(clone(change));
```
Comment (akosyakov) [inline packages/preferences-api/src/common/preference-service.ts]: It means there will be only one copy for all listeners? I am not sure how to handle it, I see 2 ways:
- instead of cloning, we do `Object.freeze` that clients will get a runtime error if they try to modify and will be forced to clone themselves
- or we don't send a value as a part of an event, clients should always use `get`

**Comment by simark** [comment]
On 1 February 2018 13:58:00 GMT-05:00, Anton Kosyakov <notifications@github.com> wrote:
>akosyakov commented on this pull request.
>
>
>
>> @@ -49,7 +50,7 @@ export class PreferenceService implements
>Disposable {
> 
>         this.resolveReady();
>         for (const change of event.changes) {
>-            this.onPreferenceChangedEmitter.fire(change);
>+            this.onPreferenceChangedEmitter.fire(clone(change));
>
>It means there will be only one copy for all listeners? I am not sure
>how to handle it, I see 2 ways:
>- instead of cloning, we do `Object.freeze` that clients will get a
>runtime error if they try to modify and will be forced to clone
>themselves
>- or we don't send a value as a part of an event, clients should always
>use `get`

I like the freeze idea, it lets us avoid unnecessary copies.  Clients can make copies if they need.

**Comment by akosyakov** [comment]
Ok, freezing is fine with me, please make sure that values are deep frozen. Also, the configuration and schema should be deep frozen when a preference proxy created.

File: packages/preferences-api/src/browser/preference-proxy.ts [hunk 5]
```diff
@@ -42,7 +42,7 @@ export function createPreferenceProxy<T extends Configuration>(preferences: Pref
                     // Fire the default preference
                     onPreferenceChangedEmitter.fire({
                         preferenceName: e.preferenceName,
-                        newValue: configuration[e.preferenceName]
+                        newValue: Object.freeze((configuration[e.preferenceName]))
```
Comment (akosyakov) [inline packages/preferences-api/src/browser/preference-proxy.ts]: no, you should freeze configuraiton and schema once and deep

File: packages/preferences-api/src/browser/preference-service.ts [hunk 6]
```diff
@@ -112,7 +112,7 @@ export class PreferenceServiceImpl implements PreferenceService, FrontendApplica
         }
         // tslint:disable-next-line:forin
         for (const preferenceName in preferenceChanges) {
-            this.onPreferenceChangedEmitter.fire(preferenceChanges[preferenceName]);
+            this.onPreferenceChangedEmitter.fire(Object.freeze(preferenceChanges[preferenceName]));
```
Comment (akosyakov) [inline packages/preferences-api/src/browser/preference-service.ts]: you should freeze values when they are created and deep, not on access

Comment (epatpol) [reply packages/preferences-api/src/browser/preference-service.ts]: I guess I can freeze them at the beginning of createPreferenceProxy, otherwise to do it before it's going to be the job of the preference contributors.

Comment (akosyakov) [reply packages/preferences-api/src/browser/preference-service.ts]: why not here: https://github.com/theia-ide/theia/pull/1175/files#diff-f719560f0cbb68164ea8381e5c8ddd3fR91?

Comment (epatpol) [reply packages/preferences-api/src/browser/preference-service.ts]: Oh yes for the actual preferences that's okay, I was talking more about configs/schemas.

**Comment by epatpol** [comment]
@akosyakov wdyt about `utils.ts`. Maybe in core?

File: packages/preferences-api/src/browser/utils.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (C) 2018 Ericsson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+export function deepFreeze(obj: any) {
```
Comment (akosyakov) [inline packages/preferences-api/src/browser/utils.ts]: please copy `deepFreeze` from https://github.com/Microsoft/vscode/blob/master/src/vs/base/common/objects.ts,

I will also prefer `objects.ts` file name, to generic `utils.ts`.

File: packages/filesystem/src/browser/filesystem-watcher.ts [hunk 8]
```diff
@@ -107,7 +106,9 @@ export class FileSystemWatcher implements Disposable {
     }
 
     protected getIgnored(): Promise<string[]> {
-        const patterns = this.preferences['files.watcherExclude'];
+        /* No need to deep copy as it's only one level deep */
+        const patterns = Object.assign({}, this.preferences['files.watcherExclude']);
```
Comment (akosyakov) [inline packages/filesystem/src/browser/filesystem-watcher.ts]: why is this change necessary?

Comment (epatpol) [reply packages/filesystem/src/browser/filesystem-watcher.ts]: Forgot filter returns a new array, removed!

File: packages/preferences-api/src/browser/objects.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (C) 2018 Ericsson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+export function deepFreeze(obj: any) {
+    Object.getOwnPropertyNames(obj).forEach(propertyName => {
```
Comment (akosyakov) [inline packages/preferences-api/src/browser/objects.ts]: please use an implementation from https://github.com/Microsoft/vscode/blob/114b6719a64566540070290a8720e89dc7ec9944/src/vs/base/common/objects.ts#L29-L47

File: packages/preferences-api/src/browser/preference-proxy.ts [hunk 10]
```diff
@@ -42,7 +45,7 @@ export function createPreferenceProxy<T extends Configuration>(preferences: Pref
                     // Fire the default preference
                     onPreferenceChangedEmitter.fire({
                         preferenceName: e.preferenceName,
-                        newValue: configuration[e.preferenceName]
+                        newValue: (configuration[e.preferenceName])
```
Comment (akosyakov) [inline packages/preferences-api/src/browser/preference-proxy.ts]: why this change?

File: packages/preferences-api/src/browser/objects.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright (C) 2018 Ericsson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+export function deepFreeze<T>(obj: T): T {
```
Comment (akosyakov) [inline packages/preferences-api/src/browser/objects.ts]: it should go to `core/common`

File: packages/preferences-api/src/browser/preference-proxy.ts (hunk 10 above)
Comment (epatpol) [reply packages/preferences-api/src/browser/preference-proxy.ts]: Reverted!
